proof-systems/src/groth16/mod.rs

```diff
@@ -208,7 +333,97 @@ impl<E: PairingEngine> ToBytes for Parameters<E>{
+        self.delta_g1.is_valid() && !self.delta_g1.is_zero() &&
+        self.delta_g2.is_valid() && !self.delta_g2.is_zero() &&
+        self.a_query.iter().filter(|&p| !p.is_valid())
+            .collect::<Vec<_>>().is_empty() &&
+        self.b_g1_query.iter().filter(|&p| !p.is_valid())
+            .collect::<Vec<_>>().is_empty() &&
+        self.b_g2_query.iter().filter(|&p| !p.is_valid())
+            .collect::<Vec<_>>().is_empty() &&
+        self.h_query.iter().filter(|&p| !p.is_valid() || p.is_zero())
+            .collect::<Vec<_>>().is_empty() &&
+        self.l_query.iter().filter(|&p| !p.is_valid() || p.is_zero())
+            .collect::<Vec<_>>().is_empty()
+    }
+}
+
+impl<E: PairingEngine> FromBytesChecked for Parameters<E> {
+    #[inline]
+    fn read_checked<R: Read>(mut reader: R) -> IoResult<Self> {
+        let vk = VerifyingKey::<E>::read_checked(&mut reader)
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
+
+        let alpha_g1 = E::G1Affine::read_checked(&mut reader)
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, format!("invalid point alpha_g1: {}", e)))
+            .and_then(|p| {
+                if p.is_zero() { return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid point alpha_g1: point at infinity")); }
+                Ok(p)
+            })?;
+
+        let beta_g1 = E::G1Affine::read_checked(&mut reader)
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, format!("invalid point beta_g1: {}", e)))
+            .and_then(|p| {
+                if p.is_zero() { return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid point beta_g1: point at infinity")); }
+                Ok(p)
+            })?;
+
+        let beta_g2 = E::G2Affine::read_checked(&mut reader)
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, format!("invalid point beta_g2: {}", e)))
+            .and_then(|p| {
+                if p.is_zero() { return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid point beta_g2: point at infinity")); }
+                Ok(p)
+            })?;
+
+        let delta_g1 = E::G1Affine::read_checked(&mut reader)
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, format!("invalid point delta_g1: {}", e)))
+            .and_then(|p| {
+                if p.is_zero() { return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid point delta_g1: point at infinity")); }
+                Ok(p)
+            })?;
+
+        let delta_g2 = E::G2Affine::read_checked(&mut reader)
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, format!("invalid point delta_g2: {}", e)))
+            .and_then(|p| {
+                if p.is_zero() { return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid point delta_g2: point at infinity")); }
+                Ok(p)
+            })?;
+
+        // NOTE: a, b_g1 and b_g2 also contains polynomials that evaluate to zero, therefore the
```

Comment:
It is extremely unlikely that one of the QAP polynomials evaluates to zero at the toxic waste point tau.
If so, I expect that such would be a danger to security due to the many factorization algorithms for polynomials over finite fields. Therefore it'd make sense to include non-zero checks, too.